### ISSUE TITLE
Stuck on infinite loop: App keeps asking for itunes password

### ISSUE THREAD

**Issue by brunolemos** [chirag04/react-native-in-app-utils#136]
So it shows the "Sign-in Required" alert.
If I press Cancel, it will close it and reopen automatically right after.
Same thing if I type the correct password and press Continue.

Is this an issue with this library, react native, ios, or what?

React Native 0.49.5
iOS 11.2

**Comment by superandrew213** [comment]
Do you try to restore purchases in the background?

**Comment by brunolemos** [comment]
@superandrew213 I call `InAppUtils.restorePurchases`

**Comment by superandrew213** [comment]
Do you call this when the user presses a button or do you call it automatically in the background to restore all the user's purchases? `InAppUtils.restorePurchases` will trigger a the "Sign-in Required" prompt when you call it.

**Comment by brunolemos** [comment]
Oh right. There are both cases:

- `InAppUtils.receiptData` triggers automatically a few seconds after the user open the app
- `InAppUtils.restorePurchases` triggers manually when the user taps on Restore

**Comment by superandrew213** [comment]
I'm triggering `InAppUtils.receiptData` in the background too but am not having any issues with it so far. However I had the same issue as you if I tried to restore the purchases in the background too.

Try disabling `InAppUtils.receiptData` and see what happens.

**Comment by brunolemos** [comment]
No luck

**Comment by brunolemos** [comment]
Even after I closed all apps the alert keeps showing to me asking for the password, even if I type it correctly... Seems to be an ios bug.

**Comment by superandrew213** [comment]
Try restarting your phone and disable `InAppUtils.receiptData`

**Comment by brunolemos** [comment]
Unfortunately didn't work. It stop asking on app open, but when I click to restore the infinite loop is back.

**Comment by brunolemos** [comment]
Maybe it's this: https://stackoverflow.com/a/24954219/2228575

Maybe we should call `finishTransaction` on the `default` switch case? 
https://github.com/chirag04/react-native-in-app-utils/blob/6bf387bb2a831f8e097547ca0fe025702d86bfa9/InAppUtils/InAppUtils.m#L61-L74

**Comment by brunolemos** [comment]
So, I added a console.log:
`NSLog(@"xxx restore request %i", queue.transactions.count);`

It's showing there are more than 40 transactions in the queue!
It definitely seems to be related to `finishTransaction` not being called somewhere.

**Comment by superandrew213** [comment]
After you clear the queue, if you press restore once, you get 40 transactions in the queue?

**Comment by chirag04** [comment]
@brunolemos if you have 40 transaction in the queue then it means you were testing something and killed/crashed the app before the transaction finished. You will have to manually clear the queue for now

**Comment by brunolemos** [comment]
> After you clear the queue, if you press restore once, you get 40 transactions in the queue?

Not 40, but a lot of them (~6?). 

> You will have to manually clear the queue for now

Do you mean by code? But what if the users get in this state? I'd like to do some fix that would work for them too 

PS there is a possibility this won't happen on production. Found some people saying this is a copy bug related to testflight

**Comment by chirag04** [comment]
> there is a possibility this won't happen on production

Yes. that's the idea. Your app won't be approved by apple if this is happening systematically.

**Comment by brunolemos** [comment]
Just another info, this is happening to me and another 2 people in the team that are not developers

**Comment by chirag04** [comment]
i would suggest adding a break point in that switch and see if it's going in a case where we don't have finishTransaction. obviously queue should be flushed before this break point to isolate the problem. you can comment out `callback` calls to let the queue flush itself first.

**Comment by jeffie** [comment]
It seems a bug for iOS11,you can try like this, This is what worked for me:
https://stackoverflow.com/questions/46257633/ios-11-is-in-app-purchase-testing-using-a-sandbox-user-keeps-asking-to-sign-in

**Comment by RomanVolkov** [comment]
when you ask for StoreKit to restore purchases it places all previous transaction ins queue. and they should be finished again too

**Comment by awchang56** [comment]
I've had this affecting my phone for months. and I found a post that said, if your restore purchases button in your app is working correctly. Run the app on the affected that you think caused the sandbox sign-in required alert, tap the restore purchases button in your app, and it should fix it once and for all, if somehow you do something to make it come back again, you can just do the same thing, hit restore purchases and it should go away again. it's like hitting the reset button. Before doing this, I had exactly 55 alerts that would pwn my phone for 45 straight seconds of tapping and it would happen once or twice a day. Now it never happens. I hope this is helpful.

[stack overflow reference](https://stackoverflow.com/a/46619824)
